processing/src/main/java/org/apache/druid/query/aggregation/SingleValueDoubleAggregatorFactory.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.query.aggregation;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonTypeName;
+import org.apache.druid.segment.BaseDoubleColumnValueSelector;
+import org.apache.druid.segment.ColumnSelectorFactory;
+import org.apache.druid.segment.column.ColumnType;
+
+import javax.annotation.Nullable;
+import java.util.Objects;
+
+@JsonTypeName("singleValueDouble")
+public class SingleValueDoubleAggregatorFactory extends SingleValueAggregatorFactory
+{
+
+  @JsonCreator
+  public SingleValueDoubleAggregatorFactory(
+      @JsonProperty("name") String name,
+      @JsonProperty("fieldName") final String fieldName
+  )
+  {
+    super(name, fieldName);
+  }
+
+  @Override
+  public Aggregator factorize(ColumnSelectorFactory metricFactory)
+  {
+    final BaseDoubleColumnValueSelector valueSelector = metricFactory.makeColumnValueSelector(getFieldName());
```

Comment:
instead of a separate aggregator factory for each type, why not call `metricFactory.getColumnCapabilities(getFieldName())` and pick the right type of aggregator.

even further, do we even need aggs/buffer aggs for each type? Given that the typed aggs all appear to be storing values as Objects instead of java numeric primitives, why not just have a single `Aggregator` and `BufferAggregator` implementation. For the `BufferAggregator`, you could pass the `ColumnType` converted from `getColumnCapabilities` and use `.getNullableStrategy()` to write and read the value from the buffer.

This seems like a lot of classes for something that is basically a value holder...

Reply:
Thanks for pointing out the NullableTypeStrategy. I always forget that such a contract exists (even in https://github.com/apache/druid/pull/15559 😞 ), and misguided the author into thinking that we don't have such a contract, and it needs to be inferred on per type basis.